### PR TITLE
nixos/qemu-vm: use cfg.host.pkgs

### PR DESCRIPTION
**Pull request by nikstur**
###### Description of changes

Hopefully fixes at least part of the issue described here: https://github.com/NixOS/nixpkgs/pull/236656#issuecomment-1596913212

I'm not sure if all these instances need to be exchanged for `cfg.host.pkgs` but it seems right to me. 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by emilazy — APPROVED** [review]
Thank you! This looks right (dunno about the `closureInfo` but I'll defer to you). I'll look into `e2fsprogs` myself soon.

**Review by RaitoBezarius — APPROVED** [review]
This looks good indeed! Thank you

**Comment by RaitoBezarius** [comment]
We need a Darwin person to confirm this is fixing darwin.builder.

**Comment by Gabriella439** [comment]
I confirmed locally that this PR fixes `darwin.builder` on `aarch64-darwin`.  Specifically, I first tested that `darwin.builder` failed locally on `master` for the reason described in https://github.com/NixOS/nixpkgs/pull/236656#issuecomment-1596913212 and then worked when I switched to this branch.

*HOWEVER*, I had to also wipe the old `nixos.qcow2` image before it worked again.

**Comment by nikstur** [comment]
@Gabriella439 thank you for testing

> _HOWEVER_, I had to also wipe the old `nixos.qcow2` image before it worked again.

Sadly, I don't think there is anything we can do about that.

**Comment by emilazy** [comment]
> We need a Darwin person to confirm this is fixing darwin.builder.

It's still broken on x86_64 because of CoreFoundation rpath hook problems breaking `e2fsprogs`. I have a workaround for that but I'm waiting for the Darwin stdenv rework to see if it's necessary to PR it.

In other words, LGTM; this is more correct even if it doesn't fix the package on all platforms immediately.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (Atemu) [inline nixos/modules/virtualisation/qemu-vm.nix:100]: If we use `cfg.host.pkgs` this often, perhaps a little `let hostPkgs = cfg.host.pkgs; in` would be cleaner.

**Review by Atemu — APPROVED** [review]
I was not able to run the created machine because it prints this error but creation works again which is what this PR's purpose is.

qemu-system-aarch64: -netdev user,id=user.0,hostfwd=tcp::22-:22,: Could not set up host forwarding rule 'tcp::22-:22'

**Comment by lilyball** [comment]
> *HOWEVER*, I had to also wipe the old `nixos.qcow2` image before it worked again.

I noticed recently that you have to do this when changing the disk space option too. I really wish the qemu-vm stuff would recognize when something changes that requires wiping the image, but that seems orthogonal to this particular issue.

**Comment by emilazy** [comment]
With https://github.com/NixOS/nixpkgs/pull/239473 maybe we can eliminate the persistent image altogether? But agreed that this PR should land as-is.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (nikstur) [reply nixos/modules/virtualisation/qemu-vm.nix]: Done.

**Comment by nikstur** [comment]
> With #239473 maybe we can eliminate the persistent image altogether?

It's a little more complicated than this. There are multiple and different kinds of persistent images used in the QEMU module. The erofs PR would only lay the groundwork to get rid of one of them (the Nix Store image built with make-disk-image.nix).

**Review by Atemu — APPROVED** [review]
LGTM.

Nit: The commit message could explain itself a bit better.

**Comment by Atemu** [comment]
Should this be backported?

**Comment by emilazy** [comment]
I think the changes that broke `darwin.builder` aren't in 23.05 so it's probably fine?

**Pull request merged**
